src/CompositeRheologies.jl

```diff
@@ -0,0 +1,794 @@
+end
+
+# define rules to nondimensionalise this 
+function nondimensionalize(MatParam::Union{Parallel,CompositeRheology}, g::GeoUnits{TYPE}) where {TYPE}
+    field_new = ();
+    field = MatParam.elements;
+    for i=1:length(MatParam.elements)
+        field_nd  = nondimensionalize(field[i], g) 
+        field_new =  tuple(field_new..., field_nd)
+    end
+    MatParam = set(MatParam, Setfield.PropertyLens{:elements}(), field_new)
+
+    return MatParam
+end
+
+function dimensionalize(MatParam::Union{Parallel,CompositeRheology}, g::GeoUnits{TYPE}) where {TYPE}
+    field_new = ();
+    field = MatParam.elements;
+    for i=1:length(MatParam.elements)
+        field_nd  = dimensionalize(field[i], g) 
+        field_new =  tuple(field_new..., field_nd)
+    end
+    MatParam = set(MatParam, Setfield.PropertyLens{:elements}(), field_new)
+
+    return MatParam
+end
+
+# Print info in the REPL
+include("CompositeRheologies_print.jl")
+
+
+
+function show(io::IO, g::AbstractComposite)
+    #println(io,"Composite rheology:   ")
+
+    # Compose a string with rheological elements, so we have an overview in the REPL
+    #str = create_rheology_string("",g)
+    str = print_rheology_matrix(g)
+    println.(str)
+    
+
+    return nothing
+end
+
+function show(io::IO, a::Parallel)
+    println(io,"Parallel:   ")  
+
+    # Compose a string with rheological elements, so we have an overview in the REPL
+    #str = create_rheology_string("",a)
+    str = print_rheology_matrix(a)
+    println.(str)
+
+    return nothing
+end
+
+# COMPUTE STRAIN RATE
+"""
+    compute_εII(v::Parallel{T,N}, τII, args; tol=1e-6, verbose=false, n=1)
+
+Computing `εII` as a function of `τII` for a Parallel elements is (usually) a nonlinear problem
+"""
+function compute_εII(
+    v::Parallel{T,N}, 
+    τII::_T, 
+    args; 
+    tol=1e-6, verbose=false, n=1
+) where {T,N,_T}
+    εII = local_iterations_τII(v, τII, args; tol=tol, verbose=verbose, n=n)
+    return εII
+end
+
+"""
+    compute_εII(v::CompositeRheology{T,N}, τII, args; tol=1e-6, verbose=false, n=1)
+
+Computing `εII` as a function of `τII` for a composite element is the sum of the individual contributions
+"""
+@generated  function compute_εII(
+    v::CompositeRheology{T,N}, 
+    τII::_T, 
+    args; 
+    tol=1e-6, verbose=false
+) where {T,_T,N}
+    quote
+        Base.@_inline_meta
+        εII = zero(_T)
+        Base.Cartesian.@nexprs $N i ->
+            εII += compute_εII(v.elements[i], τII, args)
+    end
+end
+
+
+# As we don't do iterations, this is the same
+function compute_εII_AD(v::CompositeRheology, τII, args; tol=1e-6, verbose=false)
+    return  compute_εII(v, τII, args)
+end
+
+# Here we do need to do iterations
+function compute_εII_AD(v::Parallel, τII, args; tol=1e-6, verbose=false)
+    return local_iterations_τII_AD(v, τII, args; tol=tol, verbose=verbose)
+end
+
+
+# COMPUTE DEVIATORIC STRESS
+function compute_τII(v::CompositeRheology{T,N,0}, εII, args; tol=1e-6, verbose=false) where {T,N}
```

Comment:
Is the type signature of `CompositeRheology` correct? It is defined as `CompositeRheology{T, N, Npar, is_parallel, τ_it, P_it, λ_it}`

Reply:
yes, but in this case we dont need the rest of the type signature. Is that a problem to not indicate it here?

Reply:
No that's fine, seems to work! I didn't know you could ommit the last 'n'-types in a signature. Good to know!